kimchi/src/snarky/traits.rs

```diff
@@ -0,0 +1,140 @@
+//! Traits.
+
+use ark_ff::PrimeField;
+
+use super::{
+    cvar::CVar,
+};
+
+/// A snarky type is a type that can be used in a circuit.
+/// It references an equivalent "out-of-circuit" type that one can use outside of the circuit.
+/// (For example, to construct private or public inputs, or a public output, to the circuit.)
+pub trait SnarkyType<F>: std::fmt::Debug + Sized
+where
+    F: PrimeField,
+{
+    /// Some 'out-of-circuit' data, which is carried as part of Self.
+    /// This data isn't encoded as CVars in the circuit, since the data may be large (e.g. a sparse merkle tree),
+    /// or may only be used by witness computations / for debugging.
+    type Auxiliary;
+
+    /// The equivalent "out-of-circuit" type.
+    /// For example, the [super::boolean::Boolean] snarky type has an out-of-circuit type of [bool].
+    type OutOfCircuit;
+
+    /// The number of field elements that this type takes.
+    const SIZE_IN_FIELD_ELEMENTS: usize;
+
+    /// Returns the circuit variables (and auxiliary data) behind this type.
+    fn to_cvars(&self) -> (Vec<CVar<F>>, Self::Auxiliary);
```

Comment:
This probably would be better as `[CVar<F>; Self::SIZE_IN_FIELD_ELEMENTS]`.

Reply:
you can't do this in latest stable last time I checked